feat(tech-insights): creates new endpoint for fact recalculation

sonikro opened the pull request:
## Hey, I just made a Pull Request!

Closes https://github.com/backstage/backstage/issues/18286

This PR adds a new endpoint in the tech-insights router to allow users to trigger fact recalculations for a single entity, instead of being forced to wait for the scheduled job to run.


#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

sonikro commented:
@Xantier , before I continue with more development and tests, I'd like to get your input about the direction of this implementation.

This is currently working, and if I call the new endpoint, it will trigger all FactRetrievers with an entityFilter for that single entity

![image](https://github.com/backstage/backstage/assets/11502789/c514ed9e-a2a8-49f8-a04c-b931a0ab9a3b)
![image](https://github.com/backstage/backstage/assets/11502789/9715f7ff-cd63-43e9-a155-45e045fb3a9b)

Any concerns with this implementation?

Xantier commented:
> This is currently working, and if I call the new endpoint, it will trigger all FactRetrievers with an entityFilter for that single entity

It might be beneficial to add a possibility to (optionallly) reference individual fact retrievers here. Some of them might be very high effort and wouldn't need to be necessarily run with a btn click/API call

sonikro commented:
Can this PR be re-opened ? I'm finding some time to wrap it up. Also made some new commits to add the filter by FactRetrieverID. Now working on unit tests

awanlin commented:
Hi @sonikro, I've re-opened it 👍

sonikro commented:
> Hi @sonikro, I've re-opened it 👍

Thank you. I'm working on some tests right now to make sure everything works. I'll making the new method receive a list of factRetrieverIds, so people can either recalculate all facts of all factretrievers (if they dont specify any ID), or just recalculate facts for some specific fact retrievers

sonikro commented:
@Xantier , can you take a look again? I added the ability to filter which factRetrievers you want to recalculate, and also added tests for the router and the new method in the DefaultEngine

sonikro commented:
I'm a little bit concerned with the change in the **RouterOptions**. Technically the **factRetrieverEngine** is already being passed to the createRouter in all the examples, because of the spread object in here:

https://github.com/backstage/backstage/blob/aed0534408b8d543f90afb01e749c696c64a2efb/packages/backend/src/plugins/techInsights.ts#L102

So, just adding the engine to the interface just works.

However, because people that are consuming the plugin can decide how they want to call the **createRouter**, they might not be doing the spread and might not be passing the factRetrieverEngine. So technically this _could_ be a breaking change....

Thoughts?

vinzscam commented on plugins/tech-insights-backend/api-report.md:
```diff
@@ -100,6 +111,7 @@ export interface RouterOptions<
 > {
   config: Config;
   factChecker?: FactChecker<CheckType, CheckResultType>;
+  factRetrieverEngine: FactRetrieverEngine;
```
yeah as you mentioned in your comment this is a breaking change.
Perhaps you can make it optional and throw some warnings if people try to use the new endpoint and `factRetrieverEngine` hasn't been passed?

vinzscam commented on plugins/tech-insights-backend/src/service/router.ts:
```diff
@@ -191,6 +198,18 @@ export async function createRouter<
 
+  router.post('/facts/:namespace/:kind/:name', async (req, res) => {
+    const { namespace, kind, name } = req.params;
+    const { factRetrieverIds }: { factRetrieverIds: string[] } = req.body;
```
let's add some validation to verify that `factRetrieverIds` is actually a `string[]`

vinzscam commented on plugins/tech-insights-backend/src/service/router.ts:
```diff
@@ -191,6 +198,18 @@ export async function createRouter<
+      namespace,
+      factRetrieverIds,
+    });
+    res.status(201).send();
```
I don't know the logic deeply so I might be wrong but according to the logic this looks like a `200` 😄

Xantier replied:
It trigger recalculation which creates new facts on the `/facts` endpoint, but does not return a response. I believe 201 is correct here
